### DISCUSSION TITLE
Is it possible to pass parameter to a form __init__ in a collection?

### DISCUSSION THREAD

**Discussion by lallulli** [jrief/django-formset#93] [Q&A] [answered]
I have a form class contained in a `FormCollection`.

My form has an `__init__` method which expects additional parameters, such as the request object. How can I pass the request object to the form, provided that the form is created by the `FormCollection` code behind the scenes? In vanilla Django, I would use the `get_form_kwargs` method of my `FormView` class view.

Thank you!

**Comment by jrief** [comment]
Does [this method](https://github.com/jrief/django-formset/blob/671744c82358ec161b994a171c10d89bdaf7654b/formset/views.py#L194) help to solve your use-case?

**Reply by boosh** [reply]
I have the same question. I see `get_collection_kwargs` allows me to pass kwargs from the view to the collection form, but there doesn't seem to be a way of passing those kwargs from the collection to individual child forms. Even adding a dedicated setter on the child form and calling it in `__init__` on the collection doesn't work because it seems the form attribute gets deleted. I.e.:

```
class WizardForm(FormCollection):

    default_renderer = FormRenderer(
        field_css_classes="mb-3",
        fieldset_css_classes="border-2 rounded p-3 mb-3",
    )
    main_form = MainForm()
    blog_post = BlogPostForm()

    def __init__(self, *args, **kwargs):
        project = kwargs.pop('project')
        super().__init__(*args, **kwargs)
        self.blog_post.set_kwargs(project=project)
```
Gives an error:

    'WizardForm' object has no attribute 'blog_post'

Also, instantiating `blog_post` in the collection's `__init__` method means the form is never rendered.

Is there a way to directly pass collection kwargs to a form constructor or something like `set_form_kwargs`?

Thanks

**Reply by boosh** [reply, marked as answer]
Ah, I've just discovered I can reach into the holder dict. This works:

```

class WizardView(FormCollectionView):
    template_name = "core/wizards/create.html"
    collection_class = WizardForm
    success_url = "/"

    def get_collection_kwargs(self):
        kwargs = super().get_collection_kwargs()
        kwargs["project"] = get_current_project(self.request)
        return kwargs


class WizardForm(FormCollection):
    default_renderer = FormRenderer(
        field_css_classes="mb-3",
        fieldset_css_classes="border-2 rounded p-3 mb-3",
    )
    main_form = MainForm()
    blog_post = BlogPostForm()

    def __init__(self, *args, **kwargs):
        project = kwargs.pop('project')
        super().__init__(*args, **kwargs)
        self.declared_holders["blog_post"].set_kwargs(project=project)
```

**Comment by nicokant** [comment]
I'm having the same problem, but it seems that the method `set_kwargs` does not exists.

Would it be possible to define the holders just by classes, instead of defining them as class instances?
This way it would be possible to provide a method `get_holder_kwargs` to inject values down in the form hierarchy in a django compatible way? Right now the library is great, but I feel there is some friction with the django forms api.

My use case is the following:
I have a bulk list of items that I need to edit/create using a `FormCollection`, the items depend on an order, but the order is not part of the form. I need to pass the order in the __init__ of each form because it's used to pre-filter some modelchoicefield queryset.

**Reply by jrief** [reply]
Could you please create an example of that `Form` and your `FormCollection`. Currently I don't understand what you want to achieve.

> Would it be possible to define the holders just by classes, instead of defining them as class instances?

The `holder` instances must be replicated with slightly different values when creating the collection siblings. Therefore I do not see a solution for this.

**Reply by nicokant** [reply]
Hi,
I managed to find a solution to that, I'm basically modifing the forms in the holder after __init__.
Here is some of the code that I used, the complete code is here: https://github.com/NINAnor/genlab_bestilling/blob/main/src/genlab_bestilling/forms.py

```python
# url: /order/<int:pk>/items/
class OrderItemQuantityUpdateView(BulkEditCollectionView):
    collection_class = ItemQuantityCollection
    model = OrderItemQuantity

   # make sure to inject the url kwargs into the collection
    def get_collection_kwargs(self):
        kwargs = super().get_collection_kwargs()
        kwargs["context"] = {"order_id": self.kwargs["pk"]}
        return kwargs

    def get_initial(self):
        collection_class = self.get_collection_class()
        queryset = self.get_queryset()
        initial = collection_class(
            context={"order_id": self.kwargs["pk"]}
        ).models_to_list(queryset)
        return initial

# forms.py
class OrderItemQuantityForm(forms.ModelForm):
    id = forms.IntegerField(required=False, widget=forms.widgets.HiddenInput)
    
    # custom method, will be called after __init__, but before rendering the form
    def reinit(self, context):
        self.order_id = context["order_id"]
        order = Order.object.get(id=self.order_id)
        # pre-filter the available options, based on a relationship 
        self.fields["item"].queryset = OrderItem.objects.filter(type=order.type)

    # inject the order_id into the save method
    def save(self, commit=True):
        obj = super().save(commit=False)
        obj.order_id = self.order_id
        if commit:
            obj.save()
            self.save_m2m()
        return obj

    class Meta:
        model = OrderItemQuantity
        fields = ("id", "item", "quantity")
        widgets = {
            "item": Selectize(search_lookup="name_icontains"),
        }


# NOTE: the collection here is just an array of model instances, all the instances are related to an order but it's not modified in this form
class ItemQuantityCollection(FormCollection):
    min_siblings = 1
    add_label = "Add item"
    items = OrderItemQuantityForm()
    default_renderer = FormRenderer(field_css_classes="mb-3")

    def __init__(self, *args, context=None, **kwargs):
        super().__init__(*args, **kwargs)
        self.context = context or {}

        for name, holder in self.declared_holders.items():
            if name == "items":
                # this is just to keep the logic in the original form
                holder.reinit(self.context)

    def retrieve_instance(self, data):
        if data := data.get("items"):
            try:
                return OrderItemQuantity.objects.get(id=data.get("id") or -1)
            except (AttributeError, OrderItemQuantity.DoesNotExist, ValueError):
                return OrderItemQuantity(
                    item_id=data.get("item"),
                    quantity=data.get("quantity"),
                )        
```
